llvm/lib/CodeGen/MIRParser/MIRParser.cpp

```diff
@@ -327,14 +332,29 @@ bool MIRParserImpl::parseMachineFunction(Module &M, MachineModuleInfo &MMI) {
                    "' isn't defined in the provided LLVM IR");
     }
   }
-  if (MMI.getMachineFunction(*F) != nullptr)
-    return error(Twine("redefinition of machine function '") + FunctionName +
-                 "'");
 
-  // Create the MachineFunction.
-  MachineFunction &MF = MMI.getOrCreateMachineFunction(*F);
-  if (initializeMachineFunction(YamlMF, MF))
-    return true;
+  if (!MAM) {
+    if (MMI.getMachineFunction(*F) != nullptr)
+      return error(Twine("redefinition of machine function '") + FunctionName +
+                   "'");
+
+    // Create the MachineFunction.
+    MachineFunction &MF = MMI.getOrCreateMachineFunction(*F);
+    if (initializeMachineFunction(YamlMF, MF))
+      return true;
+  } else {
+    auto &FAM =
```

Comment:
can we directly pass `FunctionAnalysisManager` to `parseMachineFunctions`?

Reply:
hmm, ok for now. as @arsenm said, it'd be nice to move completely away from MMI, but there are a lot of uses right now. although spot checking a few, a lot of them are easily removed, e.g. getting `MF.getMMI().getContext()` -> `MF.getContext()`

Reply:
could you file a bug for removing MMI?